consensus/consensusfsm/fsm.go

```diff
@@ -385,54 +393,46 @@ func (m *ConsensusFSM) calibrate(evt fsm.Event) (fsm.State, error) {
 }
 
 func (m *ConsensusFSM) prepare(_ fsm.Event) (fsm.State, error) {
-	isDelegate, proposal, isProposer, locked, preCommitEndorsement, delay, err := m.ctx.Prepare()
-	switch {
-	case err != nil:
+	if err := m.ctx.Prepare(); err != nil {
 		m.ctx.Logger().Error("Error during prepare", zap.Error(err))
-		fallthrough
-	case !isDelegate:
-		return m.BackToPrepare(delay)
+		return m.BackToPrepare(0)
+	}
+	if !m.ctx.IsDelegate() {
+		return m.BackToPrepare(0)
+	}
+	proposal, err := m.ctx.Proposal()
+	if err != nil {
+		m.ctx.Logger().Error("failed to generate block proposal", zap.Error(err))
+		return m.BackToPrepare(0)
+	}
+	m.ctx.Logger().Info("Start a new round")
+	overtime := m.ctx.WaitUntil()
+	if proposal != nil {
+		m.ctx.Broadcast(proposal)
+		m.ProduceReceiveBlockEvent(proposal)
 	}
-	m.ctx.Logger().Info("Start a new round", zap.Duration("delay", delay))
 	ttl := m.cfg.AcceptBlockTTL
-	if delay > 0 {
-		time.Sleep(delay)
-	} else {
-		ttl -= delay
+	if overtime > 0 {
+		ttl -= overtime
 	}
 	// Setup timeouts
-	if preCommitEndorsement != nil {
+	if preCommitEndorsement := m.ctx.PreCommitEndorsement(); preCommitEndorsement != nil {
 		cEvt := m.ctx.NewConsensusEvent(eBroadcastPreCommitEndorsement, preCommitEndorsement)
 		m.produce(cEvt, ttl)
 		ttl += m.cfg.AcceptProposalEndorsementTTL
 		m.produce(cEvt, ttl)
 		ttl += m.cfg.AcceptLockEndorsementTTL
 		m.produce(cEvt, ttl)
-	} else {
-		m.produceConsensusEvent(eFailedToReceiveBlock, ttl)
-		ttl += m.cfg.AcceptProposalEndorsementTTL
-		m.produceConsensusEvent(eStopReceivingProposalEndorsement, ttl)
-		ttl += m.cfg.AcceptLockEndorsementTTL
-		m.produceConsensusEvent(eStopReceivingLockEndorsement, ttl)
-	}
-	ttl += m.cfg.CommitTTL
-	m.produceConsensusEvent(ePrepare, ttl)
-	if isProposer {
-		if proposal == nil {
-			return sPrepare, errors.New("no proposal")
-		}
-		m.ctx.Broadcast(proposal)
-		m.ProduceReceiveBlockEvent(proposal)
-	}
-	if locked {
-		if preCommitEndorsement != nil {
-			return sAcceptPreCommitEndorsement, nil
-		}
-		if proposal == nil {
-			return sPrepare, errors.New("no proposal")
-		}
-		m.ProduceReceiveBlockEvent(proposal)
+		ttl += m.cfg.CommitTTL
+		m.produceConsensusEvent(eStopReceivingPreCommitEndorsement, ttl)
+
+		return sAcceptPreCommitEndorsement, nil
 	}
+	m.produceConsensusEvent(eFailedToReceiveBlock, ttl)
+	ttl += m.cfg.AcceptProposalEndorsementTTL
+	m.produceConsensusEvent(eStopReceivingProposalEndorsement, ttl)
+	ttl += m.cfg.AcceptLockEndorsementTTL
+	m.produceConsensusEvent(eStopReceivingLockEndorsement, ttl)
```

Comment:
do we need line 426, 427 here?
from old code, m.cfg.CommitTTL is needed for both if{} and else{}

Reply:
Maybe we need, especially in the PreCommit status. Let me revise it.